### PR TITLE
Hypersurface II

### PR DESCRIPTION
**Pull request by bmxam**
Second intermediate PR for the development of hypersurface capabilities. In this PR:
* refactor the file `function_space` in the doc
* export `Bcube.get_nodes`
* export a new operator, `∇ₛ` corresponding to the tangential gradient
* the `Gradient` structure is now typed by a `GradientStyle` to dispatch on the classical gradient or the tangential gradient
* all the math implementations of gradient stuff have been moved to `ref2phys.jl` instead of having some parts in `gradient.jl` and some parts in `ref2phys.jl`
* new low-level gradient function names to indicate if we are differentiating a shape function or a "regular" function; and if we are differentiating with respect to the ref or phys coordinates
* refactor `projection.jl` : functions `var_on_vertices` and `var_on_centers` can now handle any `LazyOperator`
* quadrature rule for `Point` (was missing somehow)
* gmsh generator of a cylinder mesh

There is still some on-going work on the subject by I think the PR is large enough...

### CODE REVIEW

File: src/feoperator/projection.jl [hunk 1]
```diff
@@ -44,50 +46,48 @@ function var_on_vertices(f::AbstractFEFunction, mesh::Mesh)
...
-    ξc = center(shape(ctype))
-    cPoint = CellPoint(ξc, cInfo, ReferenceDomain())
+    # Evaluate the function on the center of the cell
+    cPoint = CellPoint(center(shape(celltype(cInfo))), cInfo, ReferenceDomain())
+    value = _f(cPoint)
 
-    # Return the type
-    return eltype(_f(cPoint))
+    # Codim and type
+    N = value isa Number ? (1,) : size(N)
```
Comment (ghislainb) [inline src/feoperator/projection.jl:70]: ` : size(value)` ?

File: src/algebra/gradient.jl [hunk 2]
```diff
@@ -119,36 +115,60 @@ function Gradient(op::AbstractLazy, cPoint::CellPoint{ReferenceDomain})
...
+    ::VolumicStyle,
 )
     cnodes = get_cellnodes(cPoint)
     ctype = get_celltype(cPoint)
     ξ = get_coord(cPoint)
     fs = get_function_space(cellFunction)
     n = Val(get_size(cellFunction))
-    MapOver(grad_shape_functionsNA(fs, n, ctype, cnodes, ξ))
+    grad = ∂λξ_∂x(fs, Val(1), ctype, cnodes, ξ)
+    MapOver(grad_shape_functions_reshaped(grad, n))
```
[the remark this answers is not in the document]
Comment (bmxam) [reply src/algebra/gradient.jl:150]: `_grad_shape_functions`

File: src/feoperator/projection.jl (hunk 1 above)
Comment (bmxam) [reply src/feoperator/projection.jl:70]: + unit test

**Comment by bmxam** [comment]
@ghislainb could you please take a look again, with your suggestions implemented?

**Review by ghislainb — COMMENTED** [review]
It's ok to merge if tutorials and examples of BcubeTutorials are ok.

**Review by ghislainb — APPROVED** [review]

**Pull request merged**